### PR TITLE
[SPARK-39284][PS] Implement Groupby.mad

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
Implement Groupby.mad

### Why are the changes needed?
to increase pandas api coverage

### Does this PR introduce _any_ user-facing change?
yes

```
In [6]: pdf = pd.DataFrame({"A": [1, 2, 2, 1, 1], "B": [3, 2, 3, 9, 0], "C": [3, 4, 13, -14, 9]})

In [7]: psdf = ps.from_pandas(pdf)

In [8]: pdf.groupby("A")[["B", "C"]].mad()
Out[8]: 
          B         C
A                    
1  3.333333  8.888889
2  0.500000  4.500000

In [9]: psdf.groupby("A")[["B", "C"]].mad()
Out[9]: 
          B         C
A                    
1  3.333333  8.888889
2  0.500000  4.500000

In [10]: pdf.B.groupby(pdf.A).mad()
Out[10]: 
A
1    3.333333
2    0.500000
Name: B, dtype: float64

In [11]: psdf.B.groupby(psdf.A).mad()
Out[11]: 
A
1    3.333333
2    0.500000
Name: B, dtype: float64


```

### How was this patch tested?
added ut

### CODE REVIEW

File: python/pyspark/pandas/groupby.py [hunk 1]
```diff
@@ -759,6 +759,99 @@ def skew(scol: Column) -> Column:
...
+            data_spark_columns=[
+                scol_for(sdf, psser._internal.data_spark_column_names[0]) for psser in agg_columns
+            ],
+            column_labels=[psser._column_label for psser in agg_columns],
+            data_fields=[psser._internal.data_fields[0] for psser in agg_columns],
+            column_label_names=self._psdf._internal.column_label_names,
+        )
+        psdf: DataFrame = DataFrame(internal)
+
+        if len(psdf._internal.column_labels) > 0:
```
Comment (zhengruifeng) [inline python/pyspark/pandas/groupby.py:823]: it is not able to directly use `self._reduce_for_stat_function` for `mad` due to `It is not allowed to use a window function inside an aggregate function`.

The implement here refer to `_reduce_for_stat_function`, but has some changes here (apply window function to compute the `mean` for each group)

**Comment by zhengruifeng** [comment]
friendly ping @HyukjinKwon @itholic @xinrong-databricks @Yikun

File: python/pyspark/pandas/groupby.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/pandas/groupby.py:823]: Yeah, it sounds reasonable to me.

**Review by itholic — APPROVED** [review]
Looks pretty good

**Comment by zhengruifeng** [comment]
https://github.com/apache/spark/pull/36547 provides two methods `_prepare_reduce` and `_prepare_return`, which can be used in this PR to simplify codes.

let us wait for https://github.com/apache/spark/pull/36547

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon I think this PR is ready

File: python/pyspark/pandas/groupby.py [hunk 2]
```diff
@@ -753,6 +753,75 @@ def skew(self) -> FrameLike:
...
+            window = Window.partitionBy(groupkey_names).rowsBetween(
+                Window.unboundedPreceding, Window.unboundedFollowing
+            )
+            new_agg_scols = {}
+            new_stat_scols = []
+            for agg_column in agg_columns:
+                agg_column_name = agg_column._internal.data_spark_column_names[0]
+                new_agg_column_name = "__tmp_agg_col_{}__".format(agg_column_name)
```
Comment (Yikun) [inline python/pyspark/pandas/groupby.py:801]:
```suggestion
                new_agg_column_name = verify_temp_column_name(
                    psdf._internal.spark_frame, "__tmp_agg_col_{}__".format(agg_column_name)
                )
```

Maybe add a `verify_temp_column_name` check

**Review by Yikun — COMMENTED** [review]
LGTM execept a tiny comment. Awesome as always!

File: python/pyspark/pandas/groupby.py [hunk 3]
```diff
@@ -805,7 +874,7 @@ def all(self, skipna: bool = True) -> FrameLike:
         5  False
         """
         groupkey_names = [SPARK_INDEX_NAME_FORMAT(i) for i in range(len(self._groupkeys))]
-        internal, sdf = self._prepare_reduce(groupkey_names)
+        internal, _, sdf = self._prepare_reduce(groupkey_names)
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/groupby.py:877]: Hm, I can't follow this change. `_prepare_reduce` seems returning internal frame and dataframe. what's the one in the middle?

File: python/pyspark/pandas/groupby.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/pandas/groupby.py:793]: Maybe lets add some comments here.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to master, thanks all!

**Pull request closed without merging**